### PR TITLE
New Rule: requireVarDeclFirst

### LINKED ISSUE #99: Add option to enforce variable declaration at the top of declaring scope
In short, the option would require that all variable declarations be made at the top of their declaring scope -- where a JS interpreter would "hoist" them.  

Previously, I had proposed this idea and implemented it for JSHint, but JSHint is moving away from styling entirely in V3.0, so it was not accepted.  Unlike JSHint, JSCS seems like the perfect place for this sort of feature!

Rather than duplicating the detailed explanation/examples, see the issue jshint/jshint#1263 and pull request jshint/jshint#1283 for additional info and previous discussions.

This will be my first contribution to jscs, so it might take me a bit to port it over.  Hopefully it won't be very different/difficult.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/2009872-add-option-to-enforce-variable-declaration-at-the-top-of-declaring-scope?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by oredi**
Enforce variable declaration at the top of declaring scope

Fixes #99

### CODE REVIEW

File: lib/rules/require-var-decl-first.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+                        commentLength += comment.range[1] - comment.range[0] + comment.whitespaceBefore.length;
+                    }
+
+                    if (node.range[0] - whiteSpaceLength - commentLength <= lastNode.range[1]) {
+                        return;
+                    }
+                }
+
+                parentNodeStack.hasNonVarDecl = true;
```
Comment (markelog) [inline lib/rules/require-var-decl-first.js]: You modifying the AST?

File: test/specs/rules/require-var-decl-first.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                { statement: 'var a;var count;while(count<10){var b=1;}const c=1;', errorCount: 2 },
+                { statement: 'var a;var count;do{var b=1;}while(count<10);const c=1;', errorCount: 2 },
+                { statement: 'var a;for(var count=0; count<10;count++){var c;}var d;', errorCount: 3 },
+                { statement: 'var a; for (var count = 0; count < 10; count++) { var c; } var d;', errorCount: 3 },
+                { statement: 'var a;\nfor (var count = 0; count < 10; count++) {\nvar c;\n}\nvar d;', errorCount: 3 }
+            ];
+
+            scenarios.forEach(function(scenario) {
+                describe('statement' + JSON.stringify(scenario.statement), function() {
+                    it('should return ' + scenario.errorCount + ' error' + (scenario.errorCount === 1 ? '' : 's'),
```
Comment (markelog) [inline test/specs/rules/require-var-decl-first.js]: Not really a fan of that kind of declaration, i understand it is much easier to define, but it is much harder to support or isolate.

Not saying you should redo it, but want to hear some opinions about it.

File: lib/rules/require-var-decl-first.js (hunk 1 above)
Comment (markelog) [inline lib/rules/require-var-decl-first.js]: Oh, i see, you don't

**Comment by markelog** [comment]
Could you add some comments, like for `while` `for` loops, or why you need to `break`/`continue`, etc.

Need some else to give "LGTM" since this logic is not obvious like for the other rules.

**Comment by markelog** [comment]
@zxqfox @mrjoelkemp @mikesherov @hzoo i would like to merge this, so it would go in to 2.0, but need a at least one "LGTM" from you

**Comment by oredi** [comment]
@markelog sorry could not look into the comments yet. I will try to put them in before you start your monday. Although while working on this I have to say the rule was more complex then most of what I have work on so far.

**Comment by markelog** [comment]
@oredi which is makes so cool :-)

File: lib/rules/require-var-decl-first.js [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            var count;
+            var comment;
+            var commentLength;
+            var parentNode;
+            var parentNodeStack;
+            var lastNode;
+            var whiteSpaceLength;
+
+            parentNode = node.parentNode;
+            while (parentNode.type !== 'Program' &&
```
Comment (mrjoelkemp) [inline lib/rules/require-var-decl-first.js]: Extract to a function so the name indicates what this does. Not clear why this is necessary.

Comment (mrjoelkemp) [inline lib/rules/require-var-decl-first.js]: Ah, i see. Youre trying to find the wrapping function of the node,  if any. Yes, a function like getFunctionContaining(node) would help clarify.

File: lib/rules/require-var-decl-first.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                parentNode.type !== 'FunctionDeclaration' &&
+                parentNode.type !== 'FunctionExpression') {
+                parentNode = parentNode.parentNode;
```
Comment (mrjoelkemp) [inline lib/rules/require-var-decl-first.js]: parentNode represents the containing function. It should be called such. This should help make the code clearer.

File: lib/rules/require-var-decl-first.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            }
+            parentNodeStack = parentNodes[parentNode.range[0]];
+
+            if (parentNodeStack.declarations.length === 0) {
+                parentNodeStack.declarations.push(node);
+            } else {
+                lastNode = parentNodeStack.declarations[parentNodeStack.declarations.length - 1];
+                parentNodeStack.declarations.push(node);
+                if (!parentNodeStack.hasNonVarDecl) {
+                    if (node.range[0] <= lastNode.range[1]) {
```
Comment (mrjoelkemp) [inline lib/rules/require-var-decl-first.js]: Can you extract to a named variable to explain what's being checked here?

File: lib/rules/require-var-decl-first.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                        return;
+                    }
+
+                    whiteSpaceLength = file.getFirstNodeToken(node).whitespaceBefore.length;
+                    commentLength = 0;
```
Comment (mrjoelkemp) [inline lib/rules/require-var-decl-first.js]: For clarity and reducing the amount of code in this callback, please extract the computation of this to a function.

File: lib/rules/require-var-decl-first.js [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                        }
+
+                        if (lastNode.range[0] >= comment.range[1]) {
+                            continue;
+                        }
+
+                        commentLength += comment.range[1] - comment.range[0] + comment.whitespaceBefore.length;
+                    }
+
+                    if (node.range[0] - whiteSpaceLength - commentLength <= lastNode.range[1]) {
```
Comment (mrjoelkemp) [inline lib/rules/require-var-decl-first.js]: Not clear what this is checking. Extract to a named variable please

File: test/specs/rules/require-var-decl-first.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+            it('should report an error for value "false"', function() {
+                assert.throws(function() {
+                    checker.configure({ requireVarDeclFirst: false });
+                });
+            });
+        });
+
+        describe('rule checks', function() {
+            var scenarios = [
```
Comment (mrjoelkemp) [inline test/specs/rules/require-var-decl-first.js]: Would be easier to have a list of valid snippets. We can assume that they'd have an errorCount of zero if they were members.

File: test/specs/rules/require-var-decl-first.js (hunk 2 above)
Comment (mrjoelkemp) [inline test/specs/rules/require-var-decl-first.js]: Agreed. The test code isn't a description of the behavior being tested. It's that which tests the behavior. The number of errors isn't important when stating which behaviors or preconditions result in a failure.

**Comment by mrjoelkemp** [comment]
Great work here @oredi! Excited for this rule. However, as my comments suggest, I worry about the maintenance of the current state of the code. Any improvements to stating the high level algorithm through function usage is a big win. 

Feel free to ping me for clarification or when the edits are made. Thanks for the great effort here.

**Comment by oredi** [comment]
@markelog  and @mrjoelkemp, I will need some time to look into the changes. I have found some bugs today as well. I will update the PR again when I can get the fixes in and hopefully make the implementation clearer

**Comment by oredi** [comment]
@markelog @mrjoelkemp

I restructured the tests and code, hopefully this version is clearer. Please let me know if further edits are required

**Comment by markelog** [comment]
hm, it will fail for the case like this -

```js
function tst(node) {
    var array, // array of class names
        ncn = node.className;
}
```

**Comment by markelog** [comment]
When it is ready, would like to add to presets - 

Crockford - http://javascript.crockford.com/code.html#variable declarations
"The var statement should be the first statement in the function body." 

Wikimedia - https://www.mediawiki.org/wiki/Manual:Coding_conventions/JavaScript#Declarations
"Variables must be declared before use and should be the first statement of the function body."

WordPess - https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#declaring-variables-with-var
"Each function should begin with a single comma-delimited var statement that declares any local variables necessary."

/cc @Krinkle, @paulschreiber

**Comment by jacksonrayhamilton** [comment]
Happy to see this rule being implemented. I'll try it out soon.

I'd like to see a test case for (ignoring) the Directive Prologue (i.e., `"use strict"` directives); see [14.1.1](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-directive-prologues-and-the-use-strict-directive).

**Comment by jacksonrayhamilton** [comment]
Also, I think this rule should be called `requireVariableDeclarationFirst`, as "declaration" is spelled correctly in 5 other rules, and `requireMultipleVarDecl` should be renamed to `requireMultipleVariableDeclarations` with the old name as an obsolete alias.

**Comment by oredi** [comment]
@markelog Thanks I missed that scenario, I have put in the fix with the update to the tests.

@jacksonrayhamilton Support for this has been added to the latest commit, now the rule allows the placement of 'use strict' as the first statement of the Program, Function Declaration or Function Expression scope. I am keeping this as a separated commit as it is a special condition. If everyone agrees I combine them into a single commit.

I am keeping the name for now but if the others agree with the name change I will do that.

**Comment by mrjoelkemp** [comment]
Great job here @oredi! The code reads much better now; hopefully you agree. Thanks so much for making the edits.

**Comment by markelog** [comment]
I predict that we will receive a ticket that would ask to allow case like this - 
```js
// Example
var i;
```

**Comment by markelog** [comment]
Thank you @oredi, nicely done

**Comment by oredi** [comment]
@markelog  Thanks and I believe the following code will not cause any issue as the code handles that condition as well. Btw shall I put in another PR to add the rule to the presets for Crockford, Wikimedia and WordPess. And if yes do I need to create an issue for it or just the PR will do?
    // Example
    var i;

@mrjoelkemp Thanks. And the truth is I like this way of coding method. It literally allows me to KISS the code. :)

**Comment by oredi** [comment]
@markelog  sorry looks like you added the presets. Thanks

**Pull request closed without merging**